### ISSUE TITLE
a scan with range does not return the expected items.

### ISSUE THREAD

**Issue by C4Phoenix** [vincent-herlemont/native_db#208]
I was looking for a way too return multiple documents that contain the same nr and thought a RangeInclusive would be a good fit like so:
```rust
// does not return documents with nr = 1
let _docs = r.scan().secondary(DocumentKey::nr)?.range(1..=1).collect::<Result<Vec<Document>,_>>()?;

// returns (all the added) documents 0,1,2,3
let _docs = r.scan().secondary((DocumentKey::nr)?.range(0..=1).collect::<Result<Vec<Document>,_>>()?;

// returns no documents
let _docs = r.scan().secondary((DocumentKey::nr)?.range(1..=2).collect::<Result<Vec<Document>,_>>()?;

// returns no documents
let _docs = r.scan().secondary((DocumentKey::nr)?.range(1..2).collect::<Result<Vec<Document>,_>>()?;

// returns no documents
let _docs = r.scan().secondary((DocumentKey::nr)?.range(1..).collect::<Result<Vec<Document>,_>>()?;

// returns (all the added) documents 0,1,2,3 like expected
let _docs = r.scan().secondary((DocumentKey::nr)?.range(0..).collect::<Result<Vec<Document>,_>>()?;

// returns (all the added) documents 0,1,2,3 like expected
let _docs = r.scan().secondary((DocumentKey::nr)?.range(..10).collect::<Result<Vec<Document>,_>>()?;

// returns (all the added) documents 0,1,2,3 (so including nr 3!)
let _docs = r.scan().secondary((DocumentKey::nr)?.range(..3).collect::<Result<Vec<Document>,_>>()?;
```
`Document` has a `#[secondary_key] pub nr:u64`  for reference.

I'll be looking into the code to see if I can figure out where it goes wrong as I'd love too use this project in a personal project of mine but I'd like your help fixing this query issue. Let me know if it is reproducible or if I'm doing something stupid. Thank you in advance.

**Comment by C4Phoenix** [comment]
created a pull request with some tests that demonstrate the problem #209. They seem to only fail for non unique indexes like the ones I was using and in different ways than described above. I looked into why this would be the case but I have not yet been able to figure this out. Redb also doesn't seem too test too much for this but I'm not at all familiar with it. I kindly ask you to take a look at it.

**Comment by vincent-herlemont** [comment]
Thank you @C4Phoenix for your example. You have pointed out a (big) problem! I will work/investigate to resolve this issue and find a solution. I will keep you informed.

**Comment by vincent-herlemont** [comment]
A fix is in progress here https://github.com/vincent-herlemont/native_db/pull/214. I still have some tests to run to ensure there are no regressions.

**Comment by vincent-herlemont** [comment]
Thanks, but unfortunately, the fix I proposed doesn't work... I'll think about a more sustainable solution.

**Comment by vincent-herlemont** [comment]
@C4Phoenix It's good, I finally have a solid solution (see [this comment](https://github.com/vincent-herlemont/native_db/pull/214#issuecomment-2308272038)). You can retest on the `main` branch. I added the tests that seemed relevant to me, but feel free to add more if you'd like.

The fix will be available in version `0.8.0` because we had to change the data representation in the database. I still have some refactoring to do, so I'm not going to release it just yet.
